### PR TITLE
2828 multi line input

### LINKED ISSUE #2828: Multi-line TextInput box?
In general I've found it very difficult to find information on Bokeh widgets in the documentation, so maybe this is already possible, but is there a way to specify a paragraph-sized TextInput box?  I'm using one for the user to be able to edit some code in a domain-specific language, which is awkward to do in a single-line box.  

If multiple lines are not possible, is there at least a way to change the size of the box?

### PR DESCRIPTION
**Pull request by mas-dse-pjmulroo**
All pull requests must have an associated issue in the issue tracker. If there
isn't one, please go open an issue describing the defect, deficiency or desired
feature. You can read more about our issue and PR processes in the
[wiki](https://github.com/bokeh/bokeh/wiki/BEP-1:-Issues-and-PRs-management).

- [ X ] issues: fixes #2828
- [ - ] tests added / passed
- [ X ] release document entry (if new feature or API change)

Didn't get a chance to write the tests yet. It will take me some time to get to them.

### CODE REVIEW

File: bokeh/models/widgets/inputs.py [hunk 1]
```diff
@@ -263,3 +263,35 @@ class DatePicker(InputWidget):
     callback = Instance(Callback, help="""
     A callback to run in the browser whenever the current date value changes.
     """)
+
+class TextareaInput(InputWidget):
+    ''' Single-line input widget.
```
Comment (bryevdv) [inline bokeh/models/widgets/inputs.py]: This description is not accurate (copy/paste error)

File: bokeh/models/widgets/inputs.py [hunk 2]
```diff
@@ -263,3 +263,35 @@ class DatePicker(InputWidget):
...
+    '''
+
+    value = String(default="", help="""
+    Initial or entered text value.
+    """)
+
+    callback = Instance(Callback, help="""
+    A callback to run in the browser whenever the user unfocuses the TextInput
+    widget by hitting Enter or clicking outside of the text box area.
+    """)
```
Comment (bryevdv) [inline bokeh/models/widgets/inputs.py]: This probably needs discussion. It's now possible to add callbacks to arbitrary properties such as `value` and I think that should be the primary mechanism going forward. We've discussed deprecating these "ad-hoc" callbacks, so I don't think they should be added to any new models. But doing that will require a slight refactor to move `callback` off the base class (another reason it's not needed here)

File: bokeh/models/widgets/inputs.py [hunk 3]
```diff
@@ -263,3 +263,35 @@ class DatePicker(InputWidget):
...
+    Column count
+    """)
+
+    rows = String(default="", help="""
+    Row count
+    """)
+
+    maxlength = String(default="", help="""
+    Max count of characters in field
+    """)
```
Comment (bryevdv) [inline bokeh/models/widgets/inputs.py]: Shouldn't `cols`, `rows` and `maxlength` be `Int` properties? Also `max_length` would be more consistent with our existing codebase standards.

File: bokehjs/src/coffee/models/widgets/textarea_input.coffee [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+export class TextareaInput extends InputWidget
+  type: "TextareaInput"
+  default_view: TextareaInputView
+
+  @define {
+      value: [ p.String, "" ]
+      placeholder: [ p.String, "" ]
+      cols: [ p.String, "" ]
+      maxlength: [ p.String, "" ]
+      rows: [ p.String, "" ]
```
Comment (bryevdv) [inline bokehjs/src/coffee/models/widgets/textarea_input.coffee]: small thing but all the other models line up the values  and the brackets, so this one should too, to be consistent

File: sphinx/source/docs/user_guide/interaction/widgets.rst [hunk 5]
```diff
@@ -166,6 +166,14 @@ A widget for collecting a line of text from a user:
 .. bokeh-plot:: docs/user_guide/examples/interaction_textinput.py
     :source-position: below
 
+TextareaInput
+~~~~~~~~~
```
Comment (bryevdv) [inline sphinx/source/docs/user_guide/interaction/widgets.rst]: We have strict sphinx options turned on, this causes an error in the build. The underline length needs to match the title length exactly.

**Comment by bryevdv** [comment]
Hi @mas-dse-pjmulroo Thanks for the PR! This seems like a perfectly reasonable addition. I've left some review comments. Most are trivial, but there is one matter to decide what to do about `callback`.

**Comment by mas-dse-pjmulroo** [comment]
@bryevdv Happy to contribute, great project. Think I got all those issues addressed, but could use another look over.

File: bokehjs/src/coffee/models/widgets/textarea_input.coffee [hunk 6]
```diff
@@ -31,9 +31,9 @@ export class TextareaInput extends InputWidget
...
   @define {
-      value: [ p.String, "" ]
+      value:       [ p.String, "" ]
       placeholder: [ p.String, "" ]
-      cols: [ p.String, "" ]
-      maxlength: [ p.String, "" ]
-      rows: [ p.String, "" ]
+      cols:        [ p.Number, 20 ] # 20 is the HTML default
+      max_length:  [ p.Number, "" ]
+      rows:        [ p.Number, "" ]
```
Comment (bryevdv) [inline bokehjs/src/coffee/models/widgets/textarea_input.coffee]: I think this should be:
```
      max_length:  [ p.Number,    ]
      rows:        [ p.Number, 2  ]
```

File: bokehjs/src/coffee/models/widgets/textarea_input_template.tsx [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+export interface TextareaInputProps {
+  id: string;
+  title: string;
+  name: string;
+  placeholder: string;
+  value: string;
+  cols: string;
+  maxlength: string;
+  rows: string;
```
Comment (bryevdv) [inline bokehjs/src/coffee/models/widgets/textarea_input_template.tsx]: these also need string -> int type updates

**Comment by bryevdv** [comment]
@mas-dse-pjmulroo awesome! I only had two more small comments. Then figure out what to do about `callback`

I'm actually surprised to see that it is on the base classes as well as all the subclasses. @mattpap can you comment? I thought that would be an error. I might suggest: removing `callback` from the `InputWidget` base classes, and also removing the call to `super` in `change_input` do you agree @mattpap ?

**Comment by mas-dse-pjmulroo** [comment]
Sorry had missed max_length correction, and had to add a bit of a hack for the height. Row count was being limited by height set in parent div. Think it should work, but not a great workaround.

**Comment by bryevdv** [comment]
There's no hurry we can iterate. @mattpap may have some suggestions (better than me) for those things you mention

**Comment by bryevdv** [comment]
@mas-dse-pjmulroo FYI this branch will need a merge or rebase on master, or future test runs will fail due to an upstream dev dependency that changed last week. 

@mattpap do you have any comments on this?

**Comment by bryevdv** [comment]
Hi @mas-dse-pjmulroo there is no only one code failure, related to some automated code quality checks:
```
__________________________________ test_files __________________________________
    @pytest.mark.quality
    def test_files():
        errors = collect_errors()
>       assert len(errors) == 0, "Code quality issues:\n%s" % "\n".join(errors)
E       AssertionError: Code quality issues:
E         File contains trailing whitespace: bokeh/models/widgets/inputs.py, line 288.
E         File contains trailing whitespace: bokeh/models/widgets/inputs.py, line 292.
E         File ends with more than 1 empty line: bokeh/models/widgets/inputs.py, line 297
E       assert 3 == 0
E        +  where 3 = len(['File contains trailing whitespace: bokeh/models/widgets/inputs.py, line 288.', 'File contains trailing whitespace: b...odels/widgets/inputs.py, line 292.', 'File ends with more than 1 empty line: bokeh/models/widgets/inputs.py, line 297'])
tests/test_code_quality.py:95: AssertionError
```
Remove the trailing whitespace and all tests should pass.

**Comment by bryevdv** [comment]
last error, I think, sphinx is picky about underline lengths, they need to match the title:

https://github.com/bokeh/bokeh/pull/6162/files#diff-01b17962a9119488009b3367ef2039b5R170

**Comment by bryevdv** [comment]
@mas-dse-pjmulroo there was a recent PR that heavily refactored our low level events system. Can you merge master, and apply the following diff?
```
diff --git a/bokehjs/src/coffee/models/widgets/textarea_input.coffee b/bokehjs/src/coffee/models/widgets/textarea_input.coffee
index 578b3b599..81a5253e5 100644
--- a/bokehjs/src/coffee/models/widgets/textarea_input.coffee
+++ b/bokehjs/src/coffee/models/widgets/textarea_input.coffee
@@ -14,7 +14,7 @@ export class TextareaInputView extends InputWidgetView
   initialize: (options) ->
     super(options)
     @render()
-    @listenTo(@model, 'change', @render)
+    @connect(@model.change, @render)

   render: () ->
     super()
@@ -36,6 +36,6 @@ export class TextareaInput extends InputWidget
       value:       [ p.String, "" ]
       placeholder: [ p.String, "" ]
       cols:        [ p.Number, 20 ] # 20 is the HTML default
-      max_length:  [ p.Number ]
-      rows:        [ p.Number, 2 ] # 20 is the HTML default
+      max_length:  [ p.Number     ]
+      rows:        [ p.Number, 2  ] # 20 is the HTML default
     }

```
I think I'm willing to let the question regarding `callback` slide for now. Just be advised that `js_on_event` is definitely preferred, and that these ad-hoc `callback` properties may be deprecated in the near future. 

@mattpap I'd still appreciate your input, I'd like to know if this plays reasonably well enough with the layout machinery or if there are any show stopping changes that are needed.

**Comment by bryevdv** [comment]
@mattpap it seems to play fairly well If I replace the text inputs in the movies app with these text areas they display like this:

screen shot 2017-05-09 at 09 05 32

I would say their width is not optimally constrained (they should span the widget box width). They do respond OK to resizing:

screen shot 2017-05-09 at 09 06 17

Except it is possible to resize under the plot:

screen shot 2017-05-09 at 09 05 58

**Comment by bryevdv** [comment]
I would not consider either of those to be show-stoppers as long as their were corresponding issues added when this PR got merged. But if there are quick solutions for them we should get them in this PR.

**Comment by bryevdv** [comment]
@mattpap I will merge this tomorrow unless there is input

**Comment by bryevdv** [comment]
@mas-dse-pjmulroo this PR wil need to have master merged in, and be updated to reflect the new signal/slots API for events. You can see details in the migration notes under "New signaling API"

https://raw.githubusercontent.com/bokeh/bokeh/master/sphinx/source/docs/releases/0.12.6.rst

**Comment by bryevdv** [comment]
@mas-dse-pjmulroo It looks like something has gone amiss in the merge/update. If it's easier to just place the small set of changes on a new PR that's fine, we can close and supersede this one. Thanks for your patience

**Comment by mas-dse-pjmulroo** [comment]
@bryevdv Think I got it worked out, but probably would be nice to clean it up

**Comment by bryevdv** [comment]
@mas-dse-pjmulroo definitely needs cleanup or new PR, it's showing 400 files changed currently. :)

**Comment by bryevdv** [comment]
@mas-dse-pjmulroo I'm not sure what occurred to munge up the PR history, would you like to just submit a new clean PR based off current master?

**Comment by mas-dse-pjmulroo** [comment]
@bryevdv yeah that was my fault. I'll have a clean version in the next couple days

**Comment by bryevdv** [comment]
@mas-dse-pjmulroo I am going to go ahead and close this PR to clean up the PR list. Please submit a new clean PR whenever you are able, thanks

**Comment by kgullikson88** [comment]
Was another PR ever made for this?

**Comment by bryevdv** [comment]
Not to my knowledge

**Pull request closed without merging**